kratos/modeler/internals/cartesian_mesh_colors.h

```diff
@@ -0,0 +1,811 @@
+        return mNodalColors[index];
+    }
+
+    double& GetElementalColor(std::size_t I, std::size_t J, std::size_t K){
+        return mElementalColors[ElementIndex(I, J, K)];
+    }
+
+    double GetElementalColor(std::size_t I, std::size_t J, std::size_t K) const {
+        return mElementalColors[ElementIndex(I, J, K)];
+    }
+
+    array_1d<double,6>& GetElementalFaceColor(std::size_t I, std::size_t J, std::size_t K){
+        return mElementalFaceColors[ElementIndex(I, J, K)];
+    }
+
+    const array_1d<double,6>& GetElementalFaceColor(std::size_t I, std::size_t J, std::size_t K) const {
+        return mElementalFaceColors[ElementIndex(I, J, K)];
+    }
+
+    std::vector<double>& GetNodalColors(){ return mNodalColors;}
+
+    std::vector<double>& GetElementalColors(){return mElementalColors;}
+
+    std::vector<array_1d<double,6>>& GetElementalFaceColors(){return mElementalFaceColors;}
+
+    Point GetPoint(std::size_t I, std::size_t J, std::size_t K) const {
+        return Point(mNodalCoordinates[0][I], mNodalCoordinates[1][J], mNodalCoordinates[2][K]);
+    }
+
+    Point GetPoint(double I, double J, double K) const {
+        // The coordinates of the point are interpolated between those of floor(I, J, K) and ceil(I, J, K)
+        std::size_t i(I), j(J), k(K);
+        std::vector<double> start{mNodalCoordinates[0][i], mNodalCoordinates[1][j], mNodalCoordinates[2][k]};
+        return Point(
+            start[0] + (I-i)*(mNodalCoordinates[0][i+1]-start[0]),
+            start[1] + (J-j)*(mNodalCoordinates[1][j+1]-start[1]),
+            start[2] + (K-k)*(mNodalCoordinates[2][k+1]-start[2]));
+    }
+
+    Point GetCenterOfElement(std::size_t I, std::size_t J, std::size_t K) const {
+        return Point(mElementCenterCoordinates[0][I], mElementCenterCoordinates[1][J], mElementCenterCoordinates[2][K]);
+    }
+
+    CartesianRay<Element::GeometryType>& GetXYRay(std::size_t I, std::size_t J){
+        return mXYRays(I,J);
+    }
+
+    array_1d<std::size_t,3> GetGridSize() const {
+        array_1d<std::size_t,3> tmp{mNodalCoordinates[0].size(), mNodalCoordinates[1].size(), mNodalCoordinates[2].size()};
+        return tmp;
+    }
+
+    void InitializeRays(array_1d< std::size_t, 3 > const& MinRayPosition, array_1d< std::size_t, 3 > const& MaxRayPosition, const std::string& EntititesToColor){
+
+        std::vector<double>* p_x_coordinates = &(mNodalCoordinates[0]);
+        std::vector<double>* p_y_coordinates = &(mNodalCoordinates[1]);
+        std::vector<double>* p_z_coordinates = &(mNodalCoordinates[2]);
+
+        if((EntititesToColor == "center_of_elements") || (EntititesToColor == "face_of_elements")){
+            p_x_coordinates = &(mElementCenterCoordinates[0]);
+            p_y_coordinates = &(mElementCenterCoordinates[1]);
+            p_z_coordinates = &(mElementCenterCoordinates[2]);
+        }
+
+
+        #pragma omp parallel for
```

Comment:
I know the parallelism will go in another PR but if you want to leave this one here it should be done with `block_for_each` or commented.

Same for all in the file.